add filter by name feature

poncem91 opened the pull request:
Commit implements feature that allows user to filter comments by full name.

poncem91 committed:
add filter by name input field

Commit adds the input field for a user to enter a name to filter by name.

poncem91 committed:
Merge branch 'Mafe_DeleteSingleComment' into Mafe_FilterName

merge change to base branch

poncem91 committed:
add filter by name feature

Commit implements feature that allows user to filter comments by full name.

poncem91 committed:
Merge branch 'master' into Mafe_FilterName

merge changes to base branch

peap commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -41,9 +41,24 @@
+        System.out.println("filter is NOT empty");
+        Query.Filter queryFilter = new Query.FilterPredicate("name", Query.FilterOperator.EQUAL, filter);
+        query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING).setFilter(queryFilter);
+    }
```
Since the filter only adds on to the base query, I would create the base query first (up through `addSort`), then apply `setFilter` to that if there is indeed a filter to add. Then you don't have to repeat the first bit of code in two branches of an if-statement.

poncem91 replied:
That'll work better, thanks! I'll change this.

poncem91 committed:
remove debugging console and println messages

peap commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -78,13 +93,13 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
     // Get the input from the form.
-    String name = request.getParameter("name");
+    String name = request.getParameter("name").toLowerCase();
```
Since this erases the capitalization within provided names (e.g., "DeAngelo" becomes "Deangelo"), I would avoid it. I think it's better to preserve the user-provided name than to be able to have case-insensitive search, which is why I think you've added it. (If Datastore supported it, some kind of case-insensitive filter or fuzzy matching on your query above would be more appropriate.)

The case of letters is one of those topics[1] in software engineering that gets complicated quickly; some languages don't have case, some might not have 1-to-1 mappings between lower and upper case characters, case might be important for meaning, etc. I know said not to trust user data in a previous PR, but... :-) Transforming the input is different than guarding against possible exceptions or malicious input when it comes to user data.

[1]: Some other complicated topics: [names](https://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/), [addresses](https://www.mjt.me.uk/posts/falsehoods-programmers-believe-about-addresses/), [time](https://infiniteundo.com/post/25326999628/falsehoods-programmers-believe-about-time). The list goes on: https://github.com/kdeldycke/awesome-falsehood

poncem91 replied:
This makes a lot of sense, I hadn't thought those type of names. The list of awesome falsehoods is great! I'll change this so I don't mess with the capitalization provided.
It's a tricky issue, I did think about adding another column in datastore for 'searchname' in all lowercase, but considering the full name has to be written anyway, it didn't seem like that much of a solution in general.

peap commented on portfolio/src/main/webapp/index.html:
```diff
@@ -161,17 +161,26 @@ <h1>Contact Me</h1>
+              <div id="display-options">
+                  <div id="maxcomments-div">
+                      <label for="maxcomments">Choose the Number of Comments to Display Below:</label>
+                      <select name="maxcomments" id="maxcomments" onchange="getComments(this.value, document.getElementById('filter-input').dataset.filterInput);">
```
When the `onChange` statement gets this long, I would probably move it to a short function in your JS file and reference that. You can find out what the expected function looks like here: https://developer.mozilla.org/en-US/docs/Web/API/GlobalEventHandlers/onchange

poncem91 replied:
Got it, will change this!

peap commented on portfolio/src/main/webapp/script.js:
```diff
@@ -190,4 +192,13 @@ function deleteComments(commentId) {
+    const maxComments = document.getElementById('maxcomments').value;
+    var filterInputNode = document.getElementById('filter-input');
+    const filterInputValue = filterInputNode.value.toLowerCase();
+    filterInputNode.setAttribute("data-filter-input", filterInputValue);
```
Are you using this data attribute for anything?

poncem91 replied:
I was using it as a parameter for `getComments()` but I just realized I could've used the element's value instead. Will make some changes to this!

peap commented on portfolio/src/main/webapp/script.js:
```diff
@@ -190,4 +192,13 @@ function deleteComments(commentId) {
         const maxComments = document.getElementById('maxcomments').value;
         getComments(maxComments)
```
JavaScript is somewhat unique among the popular languages in that you can omit positional arguments from function calls, like is happening here for the new `filterInput` argument. I suggest triggering this scenario and watching the Network tab in your browser console to see what happens. (If you haven't used it yet, here's a shortish video about it: https://www.youtube.com/watch?v=nl8iKlo2NeM)

poncem91 replied:
Ooo, I didn't know that and had just missed changing that line before. I'll make some modifications now that I know this is possible in javascript, thanks!

peap reviewed (changes requested):
Nice!

poncem91 committed:
address requested changes

The user name input capitalization is no longer changed when storing it in datastore, getMaxComments() and getFilter() helper functions were added, the filter-input value was used instead of using data- attribute, the java servlet was updated to consider undefined filter parameters, and the query in the doGet() function was initialized outside of the if-statement to avoid repetitive code.

peap commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -42,23 +42,16 @@
-        query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
-    } else {
-        System.out.println("filter is NOT empty");
+    if (!filter.equals("undefined") && !filter.isEmpty()) {
```
I'd prefer to not send "undefined" as an expected value for the filter; rather, check in the JS whether `filter === undefined`, and send an empty string, if so. It's better to not use special values of the same type as normal values; maybe somebody has an online handle of "undefined" and wants to use that to leave a comment.

poncem91 replied:
That makes sense, I'll change this!

poncem91 committed:
check for undefined filterInput in JS

Commit checks whether there was an undefined filterInput argument in the getComments() function in JS. If undefined, filterInput is turned into an empty string which is then interpreted in the Java servlet as not applying any filters.
